StateMachine: balancing_{debit,credit}

sentientwaffle opened the pull request:
Add `Transfer` flags:
- `balancing_debit`
- `balancing_credit`

When set during account creation, the transfer's `amount` will be reduced to the highest nonzero amount that fulfills the debit/credit accounts' invariants (respectively).

Flag compatibility:
- The flags can be used together or apart.
- They are compatible with each other, as well as `linked` and `pending`.
- ~They are presently incompatible with `post_pending` and `void_pending` (see "open questions" below).~ https://github.com/tigerbeetledb/tigerbeetle/pull/449#issuecomment-1442780421

In my first-pass implementation I created two new error codes: `at_most_no_debits` and `at_most_no_credits`, returned when the respective flag is set but no money could be moved. But I realized that by removing them and using the existing error codes conveys additional information (i.e. which constraint is the limiting factor).

Open questions:
- ~Naming improvements?~
- ~How should this flag interact with post/void transfers? (Currently it panics with a "TODO"). Either:~
  - (We could just implement this as "if > pending amount, send pending amount", but since we already can send the full amount via post amount=0 there isn't really any reason to...)
  - ~Return a new error code.~
  - ~If we support partial post/void of pending transfers, then we can extend post/void to support `balancing_debit`/`balancing_credit`.~
- ~How should this flag interact with `exists_with_different_amount`?~
  - We could expect an exact-amount match, and return `exists_with_different_amount` otherwise. This is simplest to implement, but is probably surprising API behavior.
  - ~When `balancing_debit` or `balancing_credit`, only return `exists_with_different_amount` when the incoming transfer's amount is _less than_ the saved transfer's amount. Otherwise return `exists`.~
    - ~(Still pretty simple to implement. Slightly more intuitive, hopefully.)~
    - ~This is a bit fuzzy since the incoming transfer's amount might actually be different than it was originally, and we would still return `exists`. - I'm leaning towards this anyway, since it is a little simpler.~
  - ~Use an extra field in the durable transfer to store its original "maximum" amount.~

## Pre-merge checklist

Performance:
* I am very sure this PR could not affect performance.

sentientwaffle committed:
SM: Transfer.flags.{debits,credits}_at_most (WIP)

Add `Transfer` flags:
- `debits_at_most`
- `credits_at_most`

When set during account creation, the transfer's `amount` will be reduced to the highest nonzero amount that fulfills the debit/credit accounts' invariants (respectively).

Flag compatibility:
- The flags can be used together or apart.
- They are compatible with each other, as well as `linked` and `pending`.
- They are presently incompatible with `post_pending` and `void_pending` (see "open questions" below).

In my first-pass implementation I created two new error codes: `at_most_no_debits` and `at_most_no_credits`, returned when the respective flag is set but no money could be moved. But I realized that by removing them and using the existing error codes conveys additional information (i.e. which constraint is the limiting factor).

Open questions:
- Naming improvements?
- How should this flag interact with post/void transfers? (Currently it panics with a "TODO"). Either:
  - (We could just implement this as "if > pending amount, send pending amount", but since we already can send the full amount via post amount=0 there isn't really any reason to...)
  - Return a new error code.
  - If we support partial post/void of pending transfers, then we can extend post/void to support `debits_at_most`/`credits_at_most`.
- How should this flag interact with `exists_with_different_amount`?
  - We could expect an exact-amount match, and return `exists_with_different_amount` otherwise. This is simplest to implement, but is probably surprising API behavior.
  - When `debits_at_most` or `credits_at_most`, only return `exists_with_different_amount` when the incoming transfer's amount is _less than_ the saved transfer's amount. Otherwise return `exists`.
    - (Still pretty simple to implement. Slightly more intuitive, hopefully.)
    - This is a bit fuzzy since the incoming transfer's amount might actually be different than it was originally, and we would still return `exists`.
    - I'm leaning towards this anyway, since it is a little simpler.
  - Use an extra field in the durable transfer to store its original "maximum" amount.

sentientwaffle committed:
Clients: Update for {debits,credits}_at_most

jorangreef commented on docs/reference/transfers.md:
```diff
@@ -364,11 +366,60 @@ Mark the transfer as a [post-pending transfer](#post-pending-transfer).
+
+`flags.debits_at_most` is incompatible with:
+- `flags.post_pending_transfer`
+- `flags.void_pending_transfer`
```
```suggestion
Mutually exclusive to the `post_pending_transfer` and `void_pending_transfer` flags since [...explain why since people will otherwise wonder and ask about this...].
```

jorangreef replied:
Also, rather than making it exclusive, why not implement this by using up to the pending transfer value?

sentientwaffle replied:
I left it deliberately unimplemented/incompatible for now, waiting on a decision on partial-resolution of pending transfers, since that would involve a different implementation here anyway.

jorangreef commented on src/state_machine.zig:
```diff
@@ -692,34 +692,59 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
+                }
+                // If amount was 0, we will need to return an overflow/exceeds error.
+                break :amount std.math.max(amount, 1);
+            };
```
This block might be better served as a function, to:
1. keep the surrounding business logic clear and high level
2. use early return instead of a temp variable.

Also, try to eliminate the overflow/exceeds error earlier, explicitly, rather than using a special case `max(1)`.

sentientwaffle replied:
> Also, try to eliminate the overflow/exceeds error earlier, explicitly, rather than using a special case max(1).

Hmm, I'm not quite sure what you're envisioning here.
My rationale for the status quo: The actual error to return is one of `overflows_debits_pending`, `overflows_debits_posted`, `overflows_credits_pending`, `overflows_credits_posted`, `overflows_debits`, `overflows_credits`, `exceeds_credits`, and `exceeds_debits`. 
The `max(1)` trick gives us all of those error checks for free, no need to reimplement them for the at_most flags.

jorangreef commented on src/state_machine.zig:
```diff
@@ -692,34 +692,59 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
+                        dr.credits_posted
+                    else
+                        std.math.maxInt(u64);
+                    const dr_available = dr_limit - dr.debits_posted - dr.debits_pending;
```
It seems surprising to have `dr_available` be the potential diff between `maxInt(u64)` and the debit balance. As an interim variable it's not meaningful.

Can we rework this calculation to rather cache `dr_balance = dr.debits_posted + dr.debits_pending` (which would be more meaningful by itself), and then do the same for `cr_balance` and then put these together?

jorangreef replied:
Names like `debit balance` and `net dr/cr balance` are accounting terms—it would be great to phrase the calculation in those terms.

sentientwaffle replied:
Maybe I'm misunderstanding the definitions of "debit balance" I've found, but wouldn't it be `debits_posted + dr.debits_pending - credits_posted`, not `debits_posted + debits_pending`? And it is only applicable when the "must not exceed" flag is set.

jorangreef replied:
The *debit balance* is only the sum of the debit side.

The *net debit balance* would be the sum of the debit side less the credit side, assuming the debit side is greater.

jorangreef replied:
Sometimes people say that an account has a *debit balance*, implying *net debit balance*.

sentientwaffle replied:
> The debit balance is only the sum of the debit side.

:+1: 

> Sometimes people say that an account has a debit balance, implying net debit balance.

Yeah, that is what confused me.

jorangreef commented on src/state_machine.zig:
```diff
@@ -1802,18 +1850,147 @@ test "create_transfers: failed linked-chains are undone within a commit" {
+        \\
+        \\ lookup_transfer T1 amount  1
+        \\ lookup_transfer T2 amount  9
+        \\ lookup_transfer T3 amount 10
```
Beautiful DSL.

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -84,7 +84,9 @@ pub const TransferFlags = packed struct {
     void_pending_transfer: bool = false,
-    padding: u12 = 0,
+    debits_at_most: bool = false,
+    credits_at_most: bool = false,
```
Since `debits` or `credits` is the qualifier, and `at_most` is the feature, it would be better to swap these around, also so that the feature name lines up.

The `at_most` could then be renamed to `closing` in the sense of a closing transfer, to give:
* `closing_debit`
* `closing_credit`

jorangreef replied:
We can then call these Closing Transfers, and the docs can explain how these work in more detail (the "at most" part), whereas `debits_at_most` is not a noun so it's hard to refer to in verbal communication.

jorangreef replied:
I chatted with @lewisdaly at the office today and he grokked what `balancing_debit/credit` might mean. Perhaps that is better than `closing_debit/credit` since it avoids potential confusion on whether the account may participate in subsequent transfers.

jorangreef commented on docs/recipes/close-account.md:
```diff
@@ -0,0 +1,30 @@
+# Close Account
+
+In accounting, a _closing entry_ zeroes a (temporary) account's balance.
```
I know that some of the Investopedia articles etc. reference a temporary account. But this was confusing for me, since I've never seen closing entries described in this way before (no account is really temporary, because you keep them around forever). It's more that some accounts are for workings. Stepping back, we can remove "temporary" here completely since it's not the main point, and rather describe this as:

> In accounting, a closing entry calculates the net debit or credit balance for an account and then credits or debits this balance respectively, to zero the account's balance and move the balance to another account.

jorangreef replied:
Also, what often happens, is you close the balance at the year end, and then you might use the account again into the next year. So the account sticks around.

jorangreef commented on docs/recipes/close-account.md:
```diff
@@ -0,0 +1,30 @@
+
+In accounting, a _closing entry_ zeroes a (temporary) account's balance.
+
+### Example
```
Great to have this worked example! 🔥

jorangreef commented on docs/recipes/close-account.md:
```diff
@@ -0,0 +1,30 @@
+| Account | Debits Pending | Debits Posted | Credits Pending | Credits Posted | Flags                            |
+| ------: | -------------: | ------------: | --------------: | -------------: | -------------------------------- |
+|     `A` |              0 |            10 |               0 |             20 | `debits_must_not_exceed_credits` |
+|     `B` |              0 |            20 |               0 |             10 | `credits_must_not_exceed_debits` |
```
Can we make this more tricky so that the accounts have different net balances? e.g. One can have 10 and the other can have 5.

jorangreef replied:
So that none of the example values repeat...

sentientwaffle committed:
StateMachine: Refactor {debits,credits}_at_most

sentientwaffle committed:
Docs: Update "Close Account" recipe

- Update "closing entry" definition via Joran.
- Make the example more interesting.

sentientwaffle committed:
Transfer: s/{debits,credits}_at_most/balancing_{debit,credit}

sentientwaffle committed:
StateMachine: balancing_{debit,credit} for post/void

- `balancing_{debit,credit}` has a different meaning for post/void transfers (compared to single-phase/pending), but similar enough that it doesn't warrant a separate flag.
- For post/void transfers, `balancing_{debit,credit}` is identical.

This flag is not particularly useful for post/void — in almost all cases, just use post/void + amount=0.

sentientwaffle commented:
Updated!

- Rename `{debits,credits}_at_least` to `balancing_{debit,credit}` (thanks for excellent naming suggestion, @jorangreef!). I used these names instead of `closing_{debit,credit}` since the feature is more general than just closing accounts.
- Refactor the `amount = { ... }` block. I know you suggested splitting it into a separate function, but unless there is something nice & reusable there that I'm not seeing, I'd prefer to keep it inline because I think it's easier to follow.
- Improve the "Close Account" example in the docs.
- I'm still a little divided wrt supporting `balancing_{debit,credit}` for post/void transfer, but don't feel strongly about it one way or the other, so it is included.
  - (The flag's meaning for post/void is different than it is for single-phase/pending.
  - In particular, the flags are interchangeable, and have nothing in particular to do with the debit or credit accounts, so the flag name is (somewhat pedantically) not really correct.
  - I can't think of any concrete+practical examples where either of these would be useful (that is not already covered by `amount=0`).
  - That said, implementing it does provide symmetry by avoiding "incompatible flags" so this may well be the lesser of two evils.)

The failing Java CI appears to be some sort of client coverage check. (@batiati can you take a look?)

sentientwaffle committed:
StateMachine: Unit test balancing_{debit,credit}'s "exists"

This turns out to be quite elegant.
If a client creates a transfer T with balancing_debit/credit, they can repeat that exact create-transfer (even within the same logical batch) and based on the result code they can determine whether the full amount was delivered, or a partial amount.
Not sure if that's actually useful but it is somehow appealing nonetheless.

kakserpom commented:
> We could expect an exact-amount match, and return exists_with_different_amount otherwise. This is simplest to implement, but is probably surprising API behavior.

If transfer ID already exists then we can make sure that the given amount is more than or equal to the existing transfer amount. You should not expect an exact match because it would ruin idempotence.

kakserpom commented on docs/reference/transfers.md:
```diff
@@ -364,11 +366,50 @@ Mark the transfer as a [post-pending transfer](#post-pending-transfer).
+less than or equal to the amount that was passed to `create_transfers`.
+
+When `flags.post_pending_transfer` or `flags.void_pending_transfer` is set, the transfer amount will
+be `min(transfer.amount, pending-amount)`.
```
`pending_amount` would look better 😁

batiati committed:
Java client: Remove generated enum from test coverage report.

sentientwaffle commented:
Thanks, @batiati!

jorangreef commented:
> I'm still a little divided wrt supporting balancing_{debit,credit} for post/void transfer

Would be great to walk and talk this later today.

sentientwaffle committed:
StateMachine: Remove balancing{debit,credit} for post/void

It wasn't useful, and was possibly misleading. Even though it was similar, it was semantically different because post/void can never trigger overflow/exceeds errors.

sentientwaffle commented:
@kakserpom 

> If transfer ID already exists then we can make sure that the given amount is more than or equal to the existing transfer amount. You should not expect an exact match because it would ruin idempotence.

When dealing with balancing debit/credit transfers, there are two amounts:
1. `A₁`: The amount in the `create_transfers` event.
2. `A₂`: The actual amount transferred.

(`A₂ ≤ A₁`.)

Because of this, `exists_with_different_amount` is tricky when paired with `balancing_debit` and/or `balancing_credit`.

Here are just a few of the different approaches we could take:
1. Return `exists_with_different_amount` if the retry amount doesn't match `A₁`.
2. Return `exists_with_different_amount` if the retry amount doesn't match `A₂`. (Implemented in this PR.)
3. Return `exists_with_different_amount` if the retry amount is greater-than-or-equal-to `A₂`. (As you described above).

Reasons for picking number 2:
- The error is more precise & literally accurate — "exists with different amount" — the created transfer's amount _is_ `A₂`, not `A₁`.
- The behavior corresponds more closely with that of non-balancing transfers. (No special case in the implementation).
- Approach 1 is precise (albeit using the `A₁` rather than `A₂`), but we don't presently store `A₁`, and it isn't clear that it would be useful.
- This does mean that retries with the original `A₁` can return `exists_with_different_amount`. I wouldn't say that this "ruins idempotence" any more than `exists` does, it is just something to be aware of when using balancing transfers.

kakserpom commented:
@sentientwaffle 

- I thought we were not going to store the requested amount (`A₁`) (which in my case will always be 2**64). 
- If we are going to end up storing it, then the approach no. 1 makes the most sense. Otherwise, the approach no. 3 is the way to go.
- Approach no. 2 is not a good option because of its non-idempotent behavior: a retry is likely to produce an error which you have to ignore. Imagine a situation where you are retrying to create a balancing transfer with the exact same transfer structure and you are bound to get an error unless `A₁` matches `A₂`, an error which you wouldn't have gotten on the first try and which you don't really care about. 

> This does mean that retries with the original A₁ can return exists_with_different_amount. I wouldn't say that this "ruins idempotence" any more than exists does, it is just something to be aware of when using balancing transfers.

Well, I ignore `exists` at the client level. I don't even need to distinguish between `ok` (lack of error) and `exists`.

sentientwaffle commented:
> I thought we were not going to store the requested amount (A₁) (which in my case will always be 2**64).

Correct; I only mentioned approach 1 by way of my explanation.

> If we are going to end up storing it, then the approach no. 1 makes the most sense. Otherwise, the approach no. 3 is the way to go.

Another issue with approach 3 is that it incorrectly implies that _any_ amount greater than `A₂` would have resulted in the same `A₁`. Suppose you have an account with $100, and create a balancing transfer to deduct $10. The account's balance is now $90. If you retry the same balancing transfer (i.e. same `id`), but with an amount of $120, then approach 3 would return `exists`, as if the original balancing transfer had tried the same thing and could only move $10 (when in fact if the original `A₁` was $120, `A₂` would be $100).

I can certainly see why, with the balancing amount `A₁` always equal to 2**64, approach 3 is best.
But this feature is more general than that. In your case you can just ignore `exists_with_different_amount` like you would `exists`.

kakserpom commented:
@sentientwaffle so what approach have you chosen in the end?

sentientwaffle commented:
> @sentientwaffle so what approach have you chosen in the end?

Approach 2, i.e. return `exists_with_different_amount` if the retry amount doesn't match `A₂`.

sentientwaffle committed:
Docs: Balancing transfer exists_with_different_amount quirks

kakserpom commented:
@sentientwaffle Approach no. 2 has an unfortunate side effect: `exists_with_different_amount` makes the whole chain (of linked transfers) fail.

sentientwaffle commented:
To be clear, it only returns `exists_with_different_amount` on the retry, not on the first attempt.

> Approach no. 2 has an unfortunate side effect: `exists_with_different_amount` makes the whole chain (of linked transfers) fail.

This is true of `exists` as well, so approaches 1, 2, and 3 all share the linked transfers behavior.

jorangreef commented on docs/recipes/close-account.md:
```diff
@@ -0,0 +1,32 @@
+|             `A` |            `C` | `2^64 - 1` |                10 | `balancing_debit`  | (close account `A`) |
+|             `C` |            `B` | `2^64 - 1` |                25 | `balancing_credit` | (close account `B`) |
+
+(Pass `maxInt(u64)` as the `Transfer.amount` so that the application does not need to know (or query) the balance prior to closing the account).
```
I think we discussed rather using `0` as the null value, like we do elsewhere, e.g. for posting transfers. This saves the developer from having to think in terms of type size minus 1.

sentientwaffle replied:
Thanks for the reminder, I forgot about that!

sentientwaffle replied:
Updated.

jorangreef commented on docs/reference/operations/create_transfers.md:
```diff
@@ -206,13 +229,21 @@ The [debit account](../transfers.md#debit_account_id) has
 The [credit account](../transfers.md#credit_account_id) has
 [`flags.credits_must_not_exceed_debits`](../transfers.md#credits_must_not_exceed_debits) set,
-but `credit_account.debits_pending + credit_account.debits_posted + transfer.amount`
+but `credit_account.credits_pending + credit_account.credits_posted + transfer.amount`
```
Good catch!

jorangreef commented on docs/reference/operations/create_transfers.md:
```diff
@@ -224,6 +255,18 @@ The transfer was not created.
 [`flags.post_pending_transfer`](../transfers.md#flagspost_pending_transfer) and
 [`flags.void_pending_transfer`](../transfers.md#flagsvoid_pending_transfer).
 
+### `cannot_balance_debit`
```
I think that `cannot_balance_debit` reads more like "cannot balance the debit side", whereas `balancing_debit` reads more like "a debit that balances the credit side".

It would be better, I think, also to bring the mutually exclusive part into the error message directly, and perhaps distinguish between the flags further. For example:

`balancing_debit_is_exclusive_with_post_pending_transfer`
`balancing_debit_is_exclusive_with_void_pending_transfer`

sentientwaffle replied:
:thinking: I was avoiding that for two reasons:
- It would require 4 new error codes. And it requires an extra 2 codes for each incompatible flag we add in the future. I'm worried we could end up in an n² situation.
- With the error naming, there isn't an obvious precedence to use in general. I'm tempted to go the other extreme: just have a catch-all `incompatible_flags` error (we could use that for the post+void combination too), and let the client sort it out (since incompatible flags are always an application/dev bug).

Alternatively, maybe a naming convention that is more consistent? These flags are in the same vein as `cannot_post_and_void_pending_transfer` and `pending_transfer_cannot_post_or_void_another`, but use different conventions. A catch-all is simplest, though.

jorangreef replied:
Agreed, also on the catch-all.

Let's do `mutually_exclusive_flags`.

jorangreef replied:
Or `flags_are_mutually_exclusive` to follow MSB order.

sentientwaffle replied:
Sounds good; and I'll rename `Account`'s `mutually_exclusive_flags` to `flags_are_mutually_exclusive` so they match.

jorangreef commented on src/state_machine.zig:
```diff
@@ -915,34 +915,60 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
+                }
+
+                // If amount was 0, we will need to return an overflow/exceeds error.
+                break :amount std.math.max(amount, 1);
```
Since this logic applies regardless of whether the `balancing` flags are set, it could inadvertently change a `0` amount to a `1`, for a non-balancing transfer.

I think it would be better to have each `balancing` branch break explicitly, eliminating the temporary `amount` variable in the outer scope, to be more tightly scoped so that this kind of bug couldn't possibly happen.

Also, do we check that a transfer cannot be both a `balancing_debit` and a `balancing_credit`?

sentientwaffle replied:
Non-balancing `amount=0` has already been handled in the `amount_must_not_be_zero` branch.

`balancing_debit` and `balancing_credit` are compatible with one another, so we intentionally don't check that.
This is also why the separate balancing branches don't break explicitly.

jorangreef replied:
> Non-balancing amount=0 has already been handled in the amount_must_not_be_zero branch.

Yes, but I would rather that we don't depend on that. Lower bar for correctness.

> `balancing_debit` and `balancing_credit` are compatible with one another, so we intentionally don't check that.
This is also why the separate balancing branches don't break explicitly.

I didn't expect that we would support this. So a transfer can both balance the `dr` account's credit balance and at the same time balance the `cr` account's debit balance? Or could you give me an example of where the user would set both flags and what we would do?

sentientwaffle replied:
> So a transfer can both balance the `dr` account's credit balance and at the same time balance the `cr` account's debit balance? Or could you give me an example of where the user would set both flags and what we would do?

Sure! My thinking:

- Setting `balancing_debit` says "transfer as much as possible, up to `amount`, without violating the debit account's limits. That is, a `balancing_debit` transfer will never return `exceeds_debits`, `overflows_debits_pending`, or `overflows_debits`, _unless_ absolutely no money could be transferred.

- Setting `balancing_credit` says "transfer as much as possible, up to `amount`, without violating the _credit_ account's limits .That is, a `balancing_credit` transfer will never return `exceeds_credits`, `overflows_credits_pending`, or `overflows_credits`, _unless_ absolutely no money could be transferred.

- Setting both balancing flags provides the union of these features — transfer as much as possible, up to `amount`, without violating either account's limits. The flags are completely orthogonal — supporting the combination requires no additional code.


Closing an account is probably the most common use case for balancing flags. Closing an account requires only a single flag (balancing debit _or_ credit, depending on which account is being closed), and an amount of `0`.

But balancing transfers are more general than just closing accounts.
For other use cases, there might be account limits on both sides of the transfer.
I don't have a specific application example use case in mind for this right now, but since we support transferring money between accounts where both accounts have limits, I think that we should correspondingly support the combined balancing flags to saturate those same limits.

jorangreef replied:
Thanks for a great explanation. Would make excellent detail for the docs.

> The flags are completely orthogonal — supporting the combination requires no additional code.

Beautiful.

jorangreef reviewed (commented):
Awesome work @sentientwaffle.

sentientwaffle committed:
StateMachine: Allow balancing amount=0

sentientwaffle committed:
StateMachine: flags_are_mutually_exclusive

- Rename `CreateAccountsResult.mutually_exclusive_flags` to `CreateAccountsResult.flags_are_mutually_exclusive`.
- Add `CreateTransferResult.flags_are_mutually_exclusive`, which replaces:
  - `cannot_post_and_void_pending_transfer`,
  - `pending_transfer_cannot_post_or_void_another`,
  - `cannot_balance_debit`, and
  - `cannot_balance_credit`.
  - It is ordered between `id_must_not_be_int_max` and `debit_account_id_must_not_be_zero`.

Change `CreateAccountsResult` code ordering::
  - Move `mutually_exclusive_flags` up between `id_must_not_be_int_max` and `ledger_must_not_be_zero`. Even though `flags` is ordered in the struct after `ledger`/`code` (not before), this will allow it to correspond better to `CreateTransferResult`.

Fix `CreateTransferResult` code ordering:
  - `pending_id_must_*` is now above `ledger` errors — not sure why it was so far down before.
  - Move `timeout_reserved_for_pending_transfer` below the `pending_id` errors. Previously it preceded them; not sure why.

sentientwaffle commented:
Updated to add `flags_are_mutually_exclusive`, which replaces 4 create-transfer result codes. This was a more extensive change than just that, though:

- Rename `CreateAccountsResult.mutually_exclusive_flags` to `CreateAccountsResult.flags_are_mutually_exclusive`.
- In both `CreateAccountsResult` and `CreateTransferResult`, reorder `flags_are_mutually_exclusive` so that it immediately follows the `id` errors. Even though `flags` follows `ledger`/`code` in the structs, it makes sense to check it first since it can change behavior that follows. (This is important in the transfer — in accounts, it is just to correspond to transfers).
- Also fix `CreateTransferResult` code ordering:
    - `pending_id_must_*` are now above `ledger` errors — not sure why they were so far down before.
    - Move `timeout_reserved_for_pending_transfer` below the `pending_id` errors. Previously it preceded them; not sure why. (This was a change to both the code and the enum).

@jorangreef Please sanity-check the result-code reordering to make sure I didn't miss something silly. I think this better reflects our result-precedence philosophy but I might have missed something. :crossed_fingers:

jorangreef commented on docs/recipes/close-account.md:
```diff
@@ -16,12 +16,12 @@ Given a set of accounts:
+|             `C` |            `B` | `0`     |                25 | `balancing_credit` | (close account `B`) |
 
-(Pass `maxInt(u64)` as the `Transfer.amount` so that the application does not need to know (or query) the balance prior to closing the account).
+(Pass `0` as the `Transfer.amount` so that the application does not need to know (or query) the balance prior to closing the account).
```
Would be great to clarify that TigerBeetle will then set `Transfer.amount` to the amount actually transferred.

jorangreef commented on src/state_machine.zig:
```diff
@@ -915,34 +917,64 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
+                    else
+                        std.math.maxInt(u64);
+                    amount = std.math.min(amount, dr_limit - dr_balance);
+                }
```
If I'm reading this right, then I think we may have a bug here, when:

1. `dedits_must_not_exceed_crebits` is not set, and
2. `amount == 0` i.e. maxInt(u64), then
3. the credit balance (`credits_posted`) will be ignored completely, so that
4. a higher amount than expected will be debited.

Instead, the basic principle with these balancing transfers (setting the further "at most" feature aside), is that the calculated amount should never exceed the net credit balance, of credits less debits (taking a conservative view of credits so as to exclude pending credits, and taking a conservative view of the balancing side so as to include pending debits).

With the "at most" cap in place, it should never exceed the max cap, but the net balance should nevertheless also be enforced. Does that make sense?

Otherwise, it would be great to explore if we can DRY up both these `balancing_debit/credit` branches, by using a single inlined function into which we pass the respective amounts and flag, twice over but with the opposite values each time, since the code is perfectly symmetrical. This could also help to reduce the bulk of this block scope.

I'm also still nervous that we rely on other code further below to check that the final balancing amount is not 0. I think we should have explicit checks closer in "space" to where we do these balancing calculations, to rule out invalid values earlier. When related checks are spread out in the code base, it's almost like another form of TOCTOU, but instead of "time", the human-confusing part is "space".

For example, what happens if the account is already balanced? It seems like we should have an error code specifically in this context of balancing transfers?

sentientwaffle replied:
I am reluctant to DRY this up with a helper function — I think "hiding" this core logic with an indirection would negate the benefit of cleanup up the block.

Is this better?

```
            const amount = amount: {
                var amount = t.amount;
                if (t.flags.balancing_debit or t.flags.balancing_credit) {
                    if (amount == 0) amount = std.math.maxInt(u64);
                } else {
                    assert(amount != 0);
                }

                if (t.flags.balancing_debit) {
                    const dr_balance = dr_mut.debits_posted + dr_mut.debits_pending;
                    amount = std.math.min(amount, dr_mut.credits_posted -| dr_balance);
                    if (amount == 0) return .already_balanced_debit_account;
                }

                if (t.flags.balancing_credit) {
                    const cr_balance = cr_mut.credits_posted + cr_mut.credits_pending;
                    amount = std.math.min(amount, cr_mut.debits_posted -| cr_balance);
                    if (amount == 0) return .already_balanced_credit_account;
                }
                break :amount amount;
            };
```

jorangreef commented:
Thanks @sentientwaffle.

> Also fix CreateTransferResult code ordering:
pending_id_must_* are now above ledger errors — not sure why they were so far down before.
Move timeout_reserved_for_pending_transfer below the pending_id errors. Previously it preceded them; not sure why. (This was a change to both the code and the enum).

Nice. Can you explain though why this is a fix? Or is this more to follow the struct field order for consistency?

Do our tests still pass with the reordering (they're supposed to catch any reordering—to enforce determinism)?

sentientwaffle committed:
StateMachine: More enum order issues

Enum order did not match the true precedence in the state machine.

All of these:

- `pending_transfer_not_found`
- `pending_transfer_not_pending`
- `pending_transfer_has_different_debit_account_id`
- `pending_transfer_has_different_credit_account_id`
- `pending_transfer_has_different_ledger`
- `pending_transfer_has_different_code`
- `exceeds_pending_transfer_amount`
- `pending_transfer_has_different_amount`
- `pending_transfer_already_posted`
- `pending_transfer_already_voided`
- `pending_transfer_expired`

are higher-priority than "exists" errors, but were ordered at the end of the enum

Also these were in the wrong order in the enum:
- `exists_with_different_user_data`
- `exists_with_different_pending_id`

Lastly, Nodejs client was not updated before.

sentientwaffle committed:
Docs: Clarify zero/nonzero balancing amount

sentientwaffle committed:
StateMachine: already_balanced_{debit,credit}_account

Always balance according to debits/credits, even if no "exceeds" account flag is set.

sentientwaffle commented:
> Nice. Can you explain though why this is a fix? Or is this more to follow the struct field order for consistency?

The "fix" was to the enum order — the StateMachine code was already correct (except inasmuch as it did not match the enum order).

In particular, many errors were below the "exists" codes, despite being higher-precedence.

`exists_with_different_pending_id` and `exists_with_different_user_data` were also swapped with one another.

> Do our tests still pass with the reordering (they're supposed to catch any reordering—to enforce determinism)?

Yes; I updated the tests where appropriate. But for the most part the tests were already correct, since they test the state machine code, not the order of enum variants.

sentientwaffle commented:
(Above issue was resolved offline; replica version was outdated.) (Rebase is unrelated.)

sentientwaffle committed:
StateMachine: Remove already_balanced_{debit,credit}_account

Rationale:
- This approach provides less detail than "already balanced" errors, but is simple, consistent, and has no footguns — i.e. only 1 code to check.
- If the user wants to distinguish "exceeds credits (unbalanced)" from "exceeds credits (balanced)" they will usually just use the "_must_not_exceed_" account flag.
- But if they don't want the "_must_not_exceed_" for some reason, there is an escape hatch:
  - "Query" whether it balanced or not by sending a linked transfer of: "balancing A -> B", "balancing B -> A", (where A is the account to query and B is a dummy limitless account), and lastly a transfer that is guaranteed to fail (so that the former two don't actually move money). Then based on the error code you can tell what kind of balance it has.

jorangreef reviewed (dismissed):
Looks good!

sentientwaffle committed:
Clients/Java: Reduce coverage

Will be fixed in separate PR.
